# A-new-pipeline-to-develop-SSR-by-ddRAD-seq-data
### This pipeline was made to develop molecular markers of microsatellites. Here, you will find the code and instructions on how to use it. 

This pipeline was developed using ddRAD-seq (double digestion restriction-site associated DNA sequence) data from Agave sp accessions and the reference genome. This code crosses these two types of information, first generating the microsatellites (SSR), then filtering only the reads containing these markers and finally filtering again only the polymorphic regions in relation to the reference genome, identifying insertions and deletions using the CIGAR alignment pattern. The design of primers for the in vitro validation of polymorphic regions can be done using the PRIMER3 platform (recommended). 

![image](https://github.com/user-attachments/assets/fd882611-bfb5-4544-ad7e-747f8bb46a4c)



